[DOCS] Document `indices` cluster stats (#50527)
elastic/elasticsearch: docs/reference/cluster/stats.asciidoc

> include::{docdir}/rest-api/common-parms.asciidoc[tag=timeoutparms]

Reviewer: This includes docs for the `?master_timeout` parameter, but the cluster stats API doesn't talk to the master so this parameter is not relevant.

Author: I replaced this with only the `timeout` parm with e1b6d2c8cc2.

[[cluster-stats-api-response-body]]
==== {api-response-body-title}

`_nodes.total`::
(integer)

> Total number of nodes in the cluster.

Reviewer: The `GET /_cluster/stats/nodes/<node_id>` form of this API call is not documented correctly - in fact it takes a [list of node filters](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster.html#cluster-nodes) similarly to many other cluster APIs:

```
GET /_cat/nodes

# 127.0.0.1 15 100 6 2.48   ilm * node-0
# 127.0.0.1 32 100 6 2.48   dil - node-4
# 127.0.0.1 22 100 6 2.48   ilm - node-2
# 127.0.0.1 17 100 6 2.48   ilm - node-1
# 127.0.0.1 12 100 6 2.48   dil - node-3
#

GET /_cluster/stats/nodes/_all?filter_path=_nodes

# {
#   "_nodes": {
#     "successful": 5,
#     "total": 5,
#     "failed": 0
#   }
# }

GET /_cluster/stats/nodes/_master?filter_path=_nodes

# {
#   "_nodes": {
#     "successful": 1,
#     "total": 1,
#     "failed": 0
#   }
# }

GET /_cluster/stats/nodes/data:true?filter_path=_nodes

# {
#   "_nodes": {
#     "successful": 2,
#     "total": 2,
#     "failed": 0
#   }
# }

GET /_cluster/stats/nodes/nonexistent?filter_path=_nodes

# {
#   "_nodes": {
#     "successful": 0,
#     "total": 0,
#     "failed": 0
#   }
# }
```

Then the `_nodes.total` field is the total number of nodes selected by the node filters, which defaults to all nodes.

Author: Thanks so much @DaveCTurner. I corrected the request format and parm as well as this response parameter with 6eb0067471e.

`_nodes.successful`::
(integer)

> Number of nodes that started succesfully.

Reviewer: This is the number of nodes that responded successfully to the stats request.

Author: Fixed with 5c5e69658da19f643b4b6dc4326dbf7a2a59230f.

`_nodes.failed`::
(integer)

> Number of nodes that failed to start.

Reviewer: Missed one! This is the number of nodes that timed out or rejected the stats request, and if nonzero we also include the reason for the failure:

```
GET /_cluster/stats?filter_path=_nodes&timeout=1s

# {
#   "_nodes": {
#     "failures": [
#       {
#         "node_id": "hy7iAfvZStOy0C78T0BfJg",
#         "caused_by": {
#           "reason": "[node-4][127.0.0.1:9304][cluster:monitor/stats[n]] request_id [14552] timed out after [1009ms]",
#           "type": "receive_timeout_transport_exception"
#         },
#         "reason": "Failed node [hy7iAfvZStOy0C78T0BfJg]",
#         "type": "failed_node_exception"
#       }
#     ],
#     "successful": 4,
#     "total": 5,
#     "failed": 1
#   }
# }
```

Author: Thanks! Fixed with 23abdd43d45.

`cluster_name`::
(string)

> Name of the cluster.

Reviewer: Link `<<cluster.name>>` here?

Author: Fixed with 6957be3c5907f4c02273c2638d98bccc7f565ca4.

`cluster_uuid`::
(string)
Unique identifier for the cluster.

`timestamp`::
(https://en.wikipedia.org/wiki/Unix_time[Unix timestamp])

> Last time the cluster statistics were refreshed.

Reviewer: We should mention that this is measured in milliseconds since other units for Unix timestamps are possible.

Author: Fixed in 848685ddfea26b5985ca214c7a44c5ee8f16805a.

`status`::
(string)
Health status of the cluster based on the state of its primary and replica

> shards. Returned values include:

Reviewer: Maybe link `<<cluster-health>>` here, and if possible can we factor the definition of the health levels into one place since they're also described under the cluster health API.

We can make a stronger statement than "Returned values _include_" - these are the only possible values for this field.

Author: I added a shared parameter and link with f02feb222d0d50baece2cc2f16f9126da30b7b71.

  `green`:::

>   All primary shard and replica shards are assigned. There are at least two

Reviewer: ```suggestion
  All shards are assigned. There are at least two
```

Author: Fixed with 3f8eaa26edb2d4a673c7e2a50c8947aef4a2967e

>   copies of all shards.

Reviewer: "at least two copies of all shards" isn't true - if an index has no replicas then there is only at most one copy of each shard but this index will report `green` health as long as all the primaries are assigned.

Author: Fixed with d0fb40f4f5b634a792dc3029ee258043797fe402.

  `yellow`:::
  All primary shards are assigned, but one or more replica shards are

>   unassigned. If a node in the cluster fails, data could be lost.

Reviewer: ```suggestion
  unassigned. If a node in the cluster fails, some data could be unavailable until that node is repaired.
```

  `red`:::

>   One or more primary shards are unassigned. This can occur briefly during

Reviewer: ```suggestion
  One or more primary shards are unassigned, so some data is unavailable. This can occur briefly during
```

Author: Updated with 4bc8ba676e7.

  cluster startup as primary shards are assigned during cluster startup. At

>   other times, this status may indicate data loss.

Reviewer: Although the last sentence is true I suggest dropping it. From time to time we see users confusing data loss with `red` health and asking us to change the definition of `red` health to mean something other than data loss since it's normally temporary (a node or two are offline).

Author: Removed with 798a69315f0.

`indices.count`::
(integer)

> Total number of indices in the cluster.

Reviewer: If you don't get stats from all nodes then this is the number of indices with shards assigned to the selected nodes.

Author: Corrrected with b5004f380aa.

`indices.shards.total`::
(integer)

> Total number of shards for all indices.

Reviewer: If you don't get stats from all nodes then this is the number of shards assigned to the selected nodes.

Author: Fixed with 05a2ee76e29.

`indices.shards.primaries`::
(integer)

> Number of primary shards.

Reviewer: If you don't get stats from all nodes then this is the number of primary shards assigned to the selected nodes.

Author: Fixed with 3dc7afe930b.

`indices.shards.replication`::
(integer)

> Number of replica shards.

Reviewer: This is the ratio of the number of replicas to the number of primaries across the selected nodes.

Author: Fixed with 5b2aa81abc1

`indices.shards.index.shards.min`::
(integer)

> Lowest historic number of shards for the cluster.

Reviewer: This is the minimum number of shards in an index, counting only the shards assigned to the selected nodes.

Author: Corrected with 95e83086325.

`indices.shards.index.shards.max`::
(integer)

> Highest historic number of shards for the cluster.

Reviewer: This is the maximum number of shards in an index, counting only the shards assigned to the selected nodes.

Author: Corrected with 95e83086325.

`indices.shards.index.shards.avg`::
(integer)

> Average number of shards for the cluster.

Reviewer: This is the mean number of shards in an index, counting only the shards assigned to the selected nodes.

Author: Corrected with 95e83086325.

`indices.shards.index.primaries.min`::
(integer)

> Lowest historic number of primary shards for the cluster.

Reviewer: This is the minimum number of primary shards in an index, counting only the shards assigned to the selected nodes.

Author: Corrected with 95e83086325.

`indices.shards.index.primaries.max`::
(integer)

> Highest historic number of primary shards for the cluster.

Reviewer: This is the maximum number of primary shards in an index, counting only the shards assigned to the selected nodes.

Author: Fixed with f4a326fbf7b.

`indices.shards.index.primaries.avg`::
(integer)

> Average number of primary shards for the cluster.

Reviewer: This is the mean number of primary shards in an index, counting only the shards assigned to the selected nodes.

Author: Fixed with 57722ee113a.

`indices.shards.index.replication.min`::
(integer)

> Lowest historic number of replica shards for the cluster.

Reviewer: This is the minimum replication factor in an index, counting only the shards assigned to the selected nodes.

Author: Fixed with a35171d000a.

`indices.shards.index.replication.max`::
(integer)

> Highest historic number of replica shards for the cluster.

Reviewer: This is the maximum replication factor in an index, counting only the shards assigned to the selected nodes.

Author: Fixed with 93b1bd2ecbf.

`indices.shards.index.replication.avg`::
(integer)

> Average number of replica shards for the cluster.

Reviewer: This is the mean replication factor in an index, counting only the shards assigned to the selected nodes.

Author: Fixed with 30c98f9576d.

`indices.docs.count`::
(integer)

> Number of non-deleted documents in the cluster.

Reviewer: If using node filters this is the number of non-deleted documents in total across all primary shards assigned to the selected nodes. If not using node filters then this is the number of non-deleted documents in all primary shards.

Author: Updated the first sentence with 25831edca67.

+
This number is based on documents in Lucene segments and may include documents
from nested fields.

`indices.docs.deleted`::
(integer)

> Number of deleted documents in the cluster.

Reviewer: If using node filters this is the number of deleted documents in total across all primary shards assigned to the selected nodes. If not using node filters then this is the number of deleted documents in all primary shards.

Author: Corrected with 25831ed.

+
This number is based on documents in Lucene segments. {es} reclaims the disk
space of deleted Lucene documents when a segment is merged.

`indices.store.size`::
(<<byte-units, byte units>>)

> Size of the indices store.

Reviewer: Total size of all shards (primaries and replicas) across all selected nodes.

Author: Fixed with 00c9f1756a9.

`indices.store.size_in_bytes`::
(integer)

> Size, in bytes, of the store.

Reviewer: Total size of all shards (primaries and replicas) across all selected nodes.

Author: Fixed with a24f8ad82fc.

`indices.fielddata.memory_size`::
(<<byte-units, byte units>>)

> Memory used for fielddata cache.

Reviewer: Total size of the memory used for field data cache on all shards (primaries and replicas) across all selected nodes.

Author: Fixed with 33d78125c80.

`indices.fielddata.memory_size_in_bytes`::
(integer)

> Memory, in bytes, used for fielddata cache.

Reviewer: Total size of the memory used for field data cache on all shards (primaries and replicas) across all selected nodes.

Author: Fixed with 33d78125c80.

`indices.fielddata.evictions`::
(integer)

> Number of fielddata evictions.

Reviewer: Total number of evictions from the field data cache on all shards (primaries and replicas) across all selected nodes.

Author: Fixed with de684758951.

`indices.query_cache.memory_size`::
(<<byte-units, byte units>>)

> Memory used for query cache.

Reviewer: Total memory used for the query cache on all shards (primaries and replicas) across all selected nodes.

Author: Fixed with 62b05d32cd6.

`indices.query_cache.memory_size_in_bytes`::
(integer)

> Memory, in bytes, used for query cache.

Reviewer: Total memory used for the query cache on all shards (primaries and replicas) across all selected nodes.

Author: Fixed with 62b05d32cd6.

`indices.query_cache.total_count`::
(integer)

> Total count of hits, misses, and cached queries in the query cache.

Reviewer: Just the number of hits and misses. Total across all shards etc. etc.

Author: Fixed with 32ed29fb35d.

`indices.query_cache.hit_count`::
(integer)

> Number of query cache hits.

Reviewer: Total across all shards etc. etc.

Author: Fixed with ebf86fc650b.

`indices.query_cache.miss_count`::
(integer)

> Number of query cache misses.

Reviewer: Total across all shards etc. etc.

Author: Fixed with ebf86fc650b.

`indices.query_cache.cache_size`::
(integer)

> Size, in bytes, of the query cache.

Reviewer: This is the number of entries _currently_ in the query cache. Total across all shards etc. etc.

Author: Fixed with 5177390c89f.

`indices.query_cache.cache_count`::
(integer)

> Count of queries in the query cache.

Reviewer: This is the number of entries that have _ever_ been added to the query cache. Total across all shards etc. etc.

Author: Fixed with 5177390c89f.

`indices.query_cache.evictions`::
(integer)

> Number of query cache evictions.

Reviewer: Total across all shards etc. etc.

Author: Fixed with c1c71304820.

`indices.completion.size_in_bytes`::
(<<byte-units, byte units>>)

> Memory used for completion.

Reviewer: Total across all shards etc. etc.

Author: Fixed with c1c71304820.

`indices.completion.size_in_bytes`::
(integer)

> Memory, in bytes, used for completion.

Reviewer: Total across all shards etc. etc.

Author: Fixed with c1c71304820.

`indices.segments.count`::
(integer)

> Number of segments.

Reviewer: Total across all shards etc. etc.

Author: Fixed with c1c71304820.

`indices.segments.memory`::
(<<byte-units, byte units>>)

> Size of segments.

Reviewer: Total amount of segments memory used across all shards etc. etc. (NB not their size -- segments exist on disk so to me "size" means size on disk).

Author: Fixed with c1c71304820.

`indices.segments.memory_in_bytes`::
(integer)

> Size, in bytes, of segments.

Reviewer: Total amount of segments memory used across all shards etc. etc. (NB not their size -- segments exist on disk so to me "size" means size on disk).

Author: Fixed with c1c71304820.

`indices.segments.terms_memory`::
(<<byte-units, byte units>>)

> Size of terms in segments.

Reviewer: Total amount of terms memory used across all shards etc. etc. (terms also kinda exist on disk)

Author: Fixed with c1c713048204642ad7671e25b36d7fd56f952df5.

`indices.segments.terms_memory_in_bytes`::
(integer)

> Size, in bytes, of terms in segments.

Reviewer: Total amount of terms memory used across all shards etc. etc. (terms also kinda exist on disk)

Author: Fixed with c1c713048204642ad7671e25b36d7fd56f952df5.

`indices.segments.stored_fields_memory`::
(<<byte-units, byte units>>)

> Size of stored fields in segments.

Reviewer: Total amount of memory used for stored fields across all shards etc. etc. (stored fields also kinda exist on disk)

Author: Fixed with 3e4b778913b.

`indices.segments.stored_fields_memory_in_bytes`::
(integer)

> Size, in bytes, of stored fields in segments.

Reviewer: Total amount of memory used for stored fields across all shards etc. etc. (stored fields also kinda exist on disk)

Author: Fixed with 3e4b778913b.

`indices.segments.term_vectors_memory`::
(<<byte-units, byte units>>)

> Size of term vectors in segments.

Reviewer: Total amount of memory used for terms vectors across all shards etc. etc. (terms vectors also kinda exist on disk)

Author: Fixed with 2f73e49b1bb.

`indices.segments.term_vectors_memory_in_bytes`::
(integer)

> Size, in bytes, of term vectors in segments.

Reviewer: Total amount of memory used for terms vectors across all shards etc. etc. (terms vectors also kinda exist on disk)

Author: Fixed with 2f73e49b1bb.

`indices.segments.norms_memory`::
(<<byte-units, byte units>>)

> Size of normalization factors in segments.

Reviewer: Total amount of memory used for normalization factors across all shards etc. etc. (normalization factors also kinda exist on disk)

Author: Corrected with 0378aafe4da.

`indices.segments.norms_memory_in_bytes`::
(integer)

> Size, in bytes, of normalization factors in segments.

Reviewer: Total amount of memory used for normalization factors across all shards etc. etc. (normalization factors also kinda exist on disk)

Author: Corrected with 0378aafe4da.

`indices.segments.points_memory`::
(<<byte-units, byte units>>)

> Size of points in segments.

Reviewer: Total amount of memory used for points across all shards etc. etc. (points also kinda exist on disk)

Author: Fixed with 21700cd0fe3.

`indices.segments.points_memory_in_bytes`::
(integer)

> Size, in bytes, of points in segments.

Reviewer: Total amount of memory used for points across all shards etc. etc. (points also kinda exist on disk)

Author: Fixed with 21700cd0fe3.

`indices.segments.doc_values_memory`::
(<<byte-units, byte units>>)

> Size of doc values in segments.

Reviewer: Total amount of memory used for doc values across all shards etc. etc. (doc values also kinda exist on disk)

Author: Fixed with ea6fc1cd68b.

`indices.segments.doc_values_memory_in_bytes`::
(integer)

> Size, in bytes, of doc values in segments.

Reviewer: Total amount of memory used for doc values across all shards etc. etc. (doc values also kinda exist on disk)

Author: Fixed with ea6fc1cd68b.

`indices.segments.index_writer_memory_in_bytes`::
(<<byte-units, byte units>>)

> Memory used by the index writer.

Reviewer: Total amount of memory used by all index writers across all shards etc. etc.

Author: Fixed with f732f26a156.

`indices.segments.index_writer_memory_in_bytes`::
(integer)

> Memory, in bytes, used by the index writer.

Reviewer: Total amount of memory used by all index writers across all shards etc. etc.

Author: Fixed with f732f26a156.

`indices.segments.version_map_memory`::
(<<byte-units, byte units>>)

> Memory used by the version map.

Reviewer: Total amount of memory used by all version maps across all shards etc. etc.

Author: Fixed with c0d2ea45c0f.

`indices.segments.version_map_memory_in_bytes`::
(integer)

> Memory, in bytes, used by the version map.

Reviewer: Total amount of memory used by all version maps across all shards etc. etc.

Author: Fixed with c0d2ea45c0f.

`indices.segments.fixed_bit_set_memory`::
(<<byte-units, byte units>>)
Memory used by fixed bit sets for nested object field types and type

> filters for <<parent-join,join>> fields.

Reviewer: Total across all shards etc. etc.

Author: Fixed with d9d2c3956ea.

`indices.segments.fixed_bit_set_memory_in_bytes`::
(integer)

> Memory, in bytes, used by fixed bit sets.

Reviewer: Total across all shards etc. etc.

Author: Fixed with d9d2c3956ea.

`indices.segments.max_unsafe_auto_id_timestamp`::
(integer)

> Timestamp of the most recent retry request.

Reviewer: Timestamp of the most recently retried indexing request.

Author: Fixed with 5d8796747b2.

`indices.segments.file_sizes.size`::
(<<byte-units, byte units>>)

> Size of the segment file.

Reviewer: I don't think we populate the `file_sizes` object in cluster stats, only in index and node stats.

Author: I added a `indices.segments.file_sizes` def that points users to the node stats API with 1965909e1ab.

`indices.segments.file_sizes.size_in_bytes`::
(integer)

> Size, in bytes, of the segment file.

Reviewer: I don't think we populate the file_sizes object in cluster stats, only in index and node stats.

Author: Removed with 1965909e1abded14a7c918b2b2f945146ed47762.

`indices.segments.file_sizes.description`::
(string)

> Description of the segment file.

Reviewer: I don't think we populate the file_sizes object in cluster stats, only in index and node stats.

Author: Removed with 1965909e1abded14a7c918b2b2f945146ed47762.
